discogs_client/models.py

```
@@ -706,6 +707,11 @@ def add_release(self, release):
         add_release_url = self.fetch('resource_url') + '/releases/{}'.format(release_id)
         self.client._post(add_release_url, None)
 
+    def delete_instance(self, instance):
+        if isinstance(instance, CollectionItemInstance):
```

The add_release method has a convenience feature so that a release can be passed as the release id but also a release instance. You could do something similar here as well with the instance_id: https://github.com/AnssiAhola/discogs_client/blob/remove_instance_from_collection_folder/discogs_client/models.py#L706

---

Yeah, like I said not the most elegant 😄

At first I tried to see if I could somehow set resource_url for the `CollectionItemInstance`
and then it would be as simple as calling `instance.delete()` on it (i think), but couldn't figure it out.

Mostly because the `CollectionFolder.releases` is a `PaginatedList` of `CollectionItemInstance`s
and I didn't see a way to pass anything (dict for example) to `PaginatedList` items

---

Ok I understand. Yes that would have been very cool but unfortunately seems to not be possible... :-(

To reflect the passing as string I only see this but I am really not sure if this makes any sense and is more convenient in some situations:

```
>>> for r in folder.releases: print(r.instance_id)
...
26575953
26575889
26576266
26576687
26576232
26575706
26575675
26576356
26575781
26576945
>>>
>>>
>>>
>>>
>>>
>>>
>>>
>>> for r in folder.releases: print(r.id)
...
231253
166005
497589
840600
478375
12277
3481
552561
53085
1347446
>>> folder.delete_release("231253", "26575953")
```

---

Or passing both id's as int, doesn't matter. Also with this idea you'd have to fiddle around with argument counts. Really not sure, probably it's a bad idea. Tell me what you think honestly! Thanks :-)

---

Maybe just accept `release`/`release_id` and `instance_id` as parameters?
As a user you would have to know those regardless if you want to know which instance you are deleting, no?

It would be unlikely that someone picks a instance with random index from the `releases` and pass that to the `remove_release` method without checking what that instance actually is.
If you know what i mean.

Or am i assuming too much 😅

---

Not sure if I understand. You mean passing a fixed args amount of 2. first releas id, second instance id. As an int.

---

Yeah release id (or Release object) and instance id like so
```python
release_id = '123'
instance_id = '1'
folder.remove_release(release_id, instance_id)
```
OR
```python
release = folder.releases[0]
folder.remove_release(release, release.instance_id)
```

---

Ok I think that's a good way. Do it like that. Also now the new name remove_release fits better.

With your very first version actually the name delete_instance and passing an instance worked well together.

Now that the name is remove_release, your last proposal seems fitting well :-)

---

Now we ask @alifhughes for a final opinion and call it a day :-) Regarding which args to pass, you prefer the first version Anssi proposed, or the last one?

---

Hey both, sorry for delay! Looks good to me, I see what you mean about trying to pass either `release` or `instance_id` into and it just handles it. But I think this implementation is fine, the nuance of the API needing `instance_id` doesn't allow it to be too easy. Some thoughts:
- Having the param as `instance` and function name as `remove_release` is a bit strange but it makes sense when you look that it has to be type of `CollectionItemInstance` so I think we could leave it
- Maybe we can invert the if so that it doesn't silently fail, and do nothing, if the param is not of instance `CollectionItemInstance`, so like:

```python
if not isinstance(instance, CollectionItemInstance):
     raise TypeError("Argument must be of type CollectionItemInstace")

instance_url = self.fetch('resource_url') + '/releases/{0}/instances/{1}'.format(instance.id, instance.instance_id)
self.client._delete(instance_url);
```

- Error message could be improved though. What do you think?

---

Thank for the feedback!
So something like

```python
def remove_release(self, instance):
    if not isinstance(instance, CollectionItemInstance):
         raise TypeError("instance must be of type CollectionItemInstace")

    instance_url = self.fetch('resource_url') + '/releases/{0}/instances/{1}'.format(instance.id, instance.instance_id)
    self.client._delete(instance_url);
```
Usage example
```python
release_instance = collection_folder.releases[0]
collection_folder.remove_release(release_instance)
```
"Real life" example
```python
for release_instance in collection_folder.releases:
    if some_criteria is True:
        collection_folder.remove_release(release_instance)
```

Agreed that the TyperError message could be friendlier 😄

---

Yes, do it @AnssiAhola and let's get this merged! Nice! I like the early exit and raise! That explains a lot to the coder already when things go wrong! The slightly changed error message you advanced from @alifhughes' version is fine now IMHO. Great teamwork guys!

---

Oops that message didn't get through yesterday ^